### PR TITLE
fix(cache): use godep to build binaries

### LINKED ISSUE #3028: cache: All the given peers are not reachable when etcd/fleet are both fine on the host
When deploying a local Vagrant cluster, `builder` got stuck in an autorestart loop. Upon investigating, I discovered that cache was choking on its etcd peers:

```console
$ deisctl status builder
● deis-builder.service - deis-builder
   Loaded: loaded (/run/fleet/units/deis-builder.service; linked-runtime)
   Active: activating (start-post) since Wed 2015-02-04 21:44:25 UTC; 23min ago
  Process: 31961 ExecStopPost=/usr/bin/docker rm -f deis-builder (code=exited, status=0/SUCCESS)
  Process: 32018 ExecStartPre=/bin/sh -c docker inspect deis-builder >/dev/null && docker rm -f deis-builder || true (code=exited, status=0/SUCCESS)
  Process: 31992 ExecStartPre=/bin/sh -c IMAGE=`/run/deis/bin/get_image /deis/builder` && docker history $IMAGE >/dev/null || docker pull $IMAGE (code=exited, status=0/SUCCESS)
  Process: 31984 ExecStartPre=/bin/sh -c docker inspect deis-builder-data >/dev/null 2>&1 || docker run --name deis-builder-data -v /var/lib/docker ubuntu-debootstrap:14.04 /bin/true (code=exited, status=0/SUCCESS)
 Main PID: 32028 (sh);         : 32029 (sh)
   CGroup: /system.slice/deis-builder.service
           ├─32028 /bin/sh -c IMAGE=`/run/deis/bin/get_image /deis/builder` && docker run --name deis-builder --rm -p 2223:22 --volumes-from=deis-builder-data -c 800 -e EXTERNAL_PORT=2223 -e HOST=$COREOS_PRIVATE_IPV4 --privileged $IMAGE
           ├─32049 docker run --name deis-builder --rm -p 2223:22 --volumes-from=deis-builder-data -c 800 -e EXTERNAL_PORT=2223 -e HOST=172.17.8.100 --privileged deis/builder:v1.3.0
           └─control
             ├─11581 sleep 1
             └─32029 /bin/sh -c echo 'Waiting for builder on 2223/tcp...' && until echo 'dummy-value' | ncat $COREOS_PRIVATE_IPV4 2223 >/dev/null 2>&1; do sleep 1; done

Feb 04 22:07:45 deis-01 sh[32028]: 2015-02-04T22:07:45Z cbc90965fd3a confd[5763]: ERROR Config check failed: exit status 1
Feb 04 22:07:45 deis-01 sh[32028]: 2015-02-04T22:07:45Z cbc90965fd3a confd[5763]: ERROR Config check failed: exit status 1
Feb 04 22:07:45 deis-01 sh[32028]: builder: waiting for confd to write initial templates...
Feb 04 22:07:46 deis-01 sh[32029]: /bin/sh: line 0: echo: write error: Broken pipe
Feb 04 22:07:50 deis-01 sh[32029]: /bin/sh: line 0: echo: write error: Broken pipe
Feb 04 22:07:50 deis-01 sh[32028]: 2015-02-04T22:07:50Z cbc90965fd3a confd[5783]: ERROR Config check failed: exit status 1
Feb 04 22:07:50 deis-01 sh[32028]: 2015-02-04T22:07:50Z cbc90965fd3a confd[5783]: ERROR Config check failed: exit status 1
Feb 04 22:07:50 deis-01 sh[32028]: 2015-02-04T22:07:50Z cbc90965fd3a confd[5783]: ERROR Config check failed: exit status 1
Feb 04 22:07:50 deis-01 sh[32028]: 2015-02-04T22:07:50Z cbc90965fd3a confd[5783]: ERROR Config check failed: exit status 1
Feb 04 22:07:50 deis-01 sh[32028]: builder: waiting for confd to write initial templates...
```

```console
$ deisctl status registry@1
● deis-registry@1.service - deis-registry
   Loaded: loaded (/run/fleet/units/deis-registry@1.service; linked-runtime)
   Active: active (running) since Wed 2015-02-04 19:42:16 UTC; 2h 25min ago
  Process: 5398 ExecStartPre=/bin/sh -c docker inspect deis-registry >/dev/null && docker rm -f deis-registry || true (code=exited, status=0/SUCCESS)
  Process: 3957 ExecStartPre=/bin/sh -c IMAGE=`/run/deis/bin/get_image /deis/registry` && docker history $IMAGE >/dev/null || docker pull $IMAGE (code=exited, status=0/SUCCESS)
 Main PID: 5414 (sh)
   CGroup: /system.slice/system-deis\x2dregistry.slice/deis-registry@1.service
           ├─5414 /bin/sh -c IMAGE=`/run/deis/bin/get_image /deis/registry` && docker run --name deis-registry --rm -p 5000:5000 -e EXTERNAL_PORT=5000 -e HOST=$COREOS_PRIVATE_IPV4 $IMAGE
           └─5431 docker run --name deis-registry --rm -p 5000:5000 -e EXTERNAL_PORT=5000 -e HOST=172.17.8.102 deis/registry:v1.3.0

Feb 04 22:07:46 deis-03 sh[5414]: 2015-02-04T22:07:46Z d985d62654ad confd[10604]: ERROR 100: Key not found (/deis/cache) [67802]
Feb 04 22:07:46 deis-03 sh[5414]: registry: waiting for confd to write initial templates...
Feb 04 22:07:51 deis-03 sh[5414]: 2015-02-04T22:07:51Z d985d62654ad confd[10611]: ERROR 100: Key not found (/deis/cache) [67849]
Feb 04 22:07:51 deis-03 sh[5414]: registry: waiting for confd to write initial templates...
Feb 04 22:07:56 deis-03 sh[5414]: 2015-02-04T22:07:56Z d985d62654ad confd[10617]: ERROR 100: Key not found (/deis/cache) [67876]
Feb 04 22:07:56 deis-03 sh[5414]: registry: waiting for confd to write initial templates...
Feb 04 22:08:01 deis-03 sh[5414]: 2015-02-04T22:08:01Z d985d62654ad confd[10623]: ERROR 100: Key not found (/deis/cache) [67911]
Feb 04 22:08:01 deis-03 sh[5414]: registry: waiting for confd to write initial templates...
Feb 04 22:08:06 deis-03 sh[5414]: 2015-02-04T22:08:06Z d985d62654ad confd[10629]: ERROR 100: Key not found (/deis/cache) [67962]
Feb 04 22:08:06 deis-03 sh[5414]: registry: waiting for confd to write initial templates...
```

```console
$ deisctl status cache
● deis-cache.service - deis-cache
   Loaded: loaded (/run/fleet/units/deis-cache.service; linked-runtime)
   Active: active (running) since Wed 2015-02-04 19:41:05 UTC; 2h 27min ago
  Process: 3209 ExecStartPre=/bin/sh -c docker inspect deis-cache >/dev/null && docker rm -f deis-cache || true (code=exited, status=0/SUCCESS)
  Process: 2479 ExecStartPre=/bin/sh -c IMAGE=`/run/deis/bin/get_image /deis/cache` && docker history $IMAGE >/dev/null || docker pull $IMAGE (code=exited, status=0/SUCCESS)
 Main PID: 3219 (sh)
   CGroup: /system.slice/deis-cache.service
           ├─3219 /bin/sh -c IMAGE=`/run/deis/bin/get_image /deis/cache` && docker run --name deis-cache --rm -p 6379:6379 -e EXTERNAL_PORT=6379 -e HOST=$COREOS_PRIVATE_IPV4 $IMAGE
           └─3237 docker run --name deis-cache --rm -p 6379:6379 -e EXTERNAL_PORT=6379 -e HOST=172.17.8.101 deis/cache:v1.3.0

Feb 04 22:07:28 deis-02 sh[3219]: 2015/02/04 22:07:28 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 22:07:28 deis-02 sh[3219]: 2015/02/04 22:07:28 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 22:07:38 deis-02 sh[3219]: 2015/02/04 22:07:38 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 22:07:38 deis-02 sh[3219]: 2015/02/04 22:07:38 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 22:07:48 deis-02 sh[3219]: 2015/02/04 22:07:48 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 22:07:48 deis-02 sh[3219]: 2015/02/04 22:07:48 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 22:07:58 deis-02 sh[3219]: 2015/02/04 22:07:58 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 22:07:58 deis-02 sh[3219]: 2015/02/04 22:07:58 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 22:08:08 deis-02 sh[3219]: 2015/02/04 22:08:08 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 22:08:08 deis-02 sh[3219]: 2015/02/04 22:08:08 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
```

Initially one would reasonably think "that's an etcd/fleet issue!" Nope - both are fine.

```console
core@deis-03 ~ $ nse deis-cache
root@e7511c7f5a4a:/app# echo $HOST
172.17.8.102
```

On that host, I can use `etcdctl` and `fleetctl` just fine.

Restarting cache doesn't resolve the issue, but it does look like the errors occur after cache starts:
```console
core@deis-03 ~ $ journalctl -fu deis-cache
-- Logs begin at Wed 2015-02-04 22:58:17 UTC. --
Feb 04 23:18:56 deis-03 sh[8169]: `-._    `-.__.-'    _.-'
Feb 04 23:18:56 deis-03 sh[8169]: `-._        _.-'
Feb 04 23:18:56 deis-03 sh[8169]: `-.__.-'
Feb 04 23:18:56 deis-03 sh[8169]: [14] 04 Feb 23:18:56.132 # Server started, Redis version 2.8.19
Feb 04 23:18:56 deis-03 sh[8169]: [14] 04 Feb 23:18:56.133 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
Feb 04 23:18:56 deis-03 sh[8169]: [14] 04 Feb 23:18:56.133 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
Feb 04 23:18:56 deis-03 sh[8169]: [14] 04 Feb 23:18:56.133 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
Feb 04 23:18:56 deis-03 sh[8169]: [14] 04 Feb 23:18:56.134 * The server is now ready to accept connections on port 6379
Feb 04 23:19:06 deis-03 sh[8169]: 2015/02/04 23:19:06 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:06 deis-03 sh[8169]: 2015/02/04 23:19:06 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:16 deis-03 sh[8169]: 2015/02/04 23:19:16 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:16 deis-03 sh[8169]: 2015/02/04 23:19:16 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:26 deis-03 sh[8169]: 2015/02/04 23:19:26 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:26 deis-03 sh[8169]: 2015/02/04 23:19:26 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:36 deis-03 sh[8169]: 2015/02/04 23:19:36 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:36 deis-03 sh[8169]: 2015/02/04 23:19:36 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:46 deis-03 sh[8169]: 2015/02/04 23:19:46 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:46 deis-03 sh[8169]: 2015/02/04 23:19:46 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:56 deis-03 sh[8169]: 2015/02/04 23:19:56 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
Feb 04 23:19:56 deis-03 sh[8169]: 2015/02/04 23:19:56 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
```

This is happening for me 50% of the time on deploys, and I know a user in IRC reported the same thing happening for him (same symptoms, cache is the culprit) within the past few days. So, I can't help but think something has changed recently.

### LINKED ISSUE #3029: Go-based bin/boot scripts don't use Godep to manage dependencies
go-etcd has a backwards-incompatible change in https://github.com/coreos/go-etcd/commit/5348c80bde6223fa53c48ef546b4802db69b1d70#diff-33c1223aac4ef741731cc8c73b4d7777L295. Components that `go get` the source when they're built are getting this bad change. 

logspout was fixed in #3007, but we need to audit the other Go-based components to see if they need to be updated.

This appears to be the root cause of #3028.

refs #3007
refs #3028

### PR DESCRIPTION
**Pull request by bacongobbler**
closes #3028
closes #3029

That should do it. Phew.

### CODE REVIEW

**Comment by carmstrong** [comment]
Thanks @bacongobbler.

The other Go-based components seem to keep everything in `/image` - should we keep that for consistency, or is cache special?

**Comment by bacongobbler** [comment]
I moved it to keep everything in `image/` when shifting around stuff. You'll notice that the root is meant for the build environment such as `/bin/test` (which we never truly did anything with, so that should be removed in a future PR) and main.go. The root Dockerfile is there to build the `redis-server` binary which we eventually throw into `image/bin`. redis.conf was moved over to image/redis.conf, too.

TL;DR it used to be the other way around, but I fixed that now :)

**Comment by mboersma** [comment]
Code LGTM. Thanks for fixing this so quickly @bacongobbler.

**Comment by gabrtv** [comment]
Code LGTM.  Let's get this merged and cut a patch release ASAP, as we're seeing this in the wild.
